include/libopencm3/efm32/common/gpio_common_hglg.h

```diff
@@ -0,0 +1,337 @@
+ */
+
+/** @cond */
+#if defined(LIBOPENCM3_GPIO_H)
+/** @endcond */
+#ifndef LIBOPENCM3_EFM32_GPIO_COMMON_HGLG_H
+#define LIBOPENCM3_EFM32_GPIO_COMMON_HGLG_H
+
+#include <libopencm3/efm32/memorymap.h>
+#include <libopencm3/cm3/common.h>
+
+#define GPIO_P(i)		(GPIO_BASE + (0x24 * (i)))
+#define GPIO_PA			GPIO_P(0)
+#define GPIO_PB			GPIO_P(1)
+#define GPIO_PC			GPIO_P(2)
+#define GPIO_PD			GPIO_P(3)
+#define GPIO_PE			GPIO_P(4)
+#define GPIO_PF			GPIO_P(5)
+
+#define GPIO_P_CTRL(port)	MMIO32((port) + 0x00)
+#define GPIO_PA_CTRL		GPIO_P_CTRL(GPIO_PA)
+#define GPIO_PB_CTRL		GPIO_P_CTRL(GPIO_PB)
+#define GPIO_PC_CTRL		GPIO_P_CTRL(GPIO_PC)
+#define GPIO_PD_CTRL		GPIO_P_CTRL(GPIO_PD)
+#define GPIO_PE_CTRL		GPIO_P_CTRL(GPIO_PE)
+#define GPIO_PF_CTRL		GPIO_P_CTRL(GPIO_PF)
+
+#define GPIO_P_CTRL_DRIVEMODE_SHIFT	(0)
+#define GPIO_P_CTRL_DRIVEMODE_MASK	(0x03 << GPIO_P_CTRL_DRIVEMODE_SHIFT)
+#define GPIO_P_CTRL_DRIVEMODE(v)	\
+	(((v) << GPIO_P_CTRL_DRIVEMODE_SHIFT) & GPIO_P_CTRL_DRIVEMODE_MASK)
+#define GPIO_P_CTRL_DRIVEMODE_STANDARD	GPIO_P_CTRL_DRIVEMODE(0)
+#define GPIO_P_CTRL_DRIVEMODE_LOWEST	GPIO_P_CTRL_DRIVEMODE(1)
+#define GPIO_P_CTRL_DRIVEMODE_HIGH	GPIO_P_CTRL_DRIVEMODE(2)
+#define GPIO_P_CTRL_DRIVEMODE_LOW	GPIO_P_CTRL_DRIVEMODE(3)
+
+/* NOTE: GPIO_MODE and GPIO_MODE_MASK is generic.
+ * it is used with both GPIO_Px_MODEL and GPIO_Px_MODEH */
+#define GPIO_P_MODE_MODEx_MASK(x)	(0x0F << (((x) & 0x7) * 4))
+/* for mode: use GPIO_MODE_* */
+#define GPIO_P_MODE_MODEx(x, mode)	\
+	(((mode) << (((x) & 0x7) * 4)) & GPIO_P_MODE_MODEx_MASK(x))
+
+#define GPIO_P_MODEL(port)	MMIO32((port) + 0x04)
+#define GPIO_PA_MODEL		GPIO_P_MODEL(GPIO_PA)
+#define GPIO_PB_MODEL		GPIO_P_MODEL(GPIO_PB)
+#define GPIO_PC_MODEL		GPIO_P_MODEL(GPIO_PC)
+#define GPIO_PD_MODEL		GPIO_P_MODEL(GPIO_PD)
+#define GPIO_PE_MODEL		GPIO_P_MODEL(GPIO_PE)
+
+#define GPIO_P_MODEL_MODEx_MASK(x)	GPIO_P_MODE_MODEx_MASK(x)
+#define GPIO_P_MODEL_MODEx(x, mode)	GPIO_P_MODE_MODEx(x, mode)
+
+#define GPIO_P_MODEL_MODE0_MASK		GPIO_P_MODEL_MODEx_MASK(0)
+#define GPIO_P_MODEL_MODE0(mode)	GPIO_P_MODEL_MODEx(0, mode)
+
+#define GPIO_P_MODEL_MODE1_MASK		GPIO_P_MODEL_MODEx_MASK(1)
+#define GPIO_P_MODEL_MODE1(mode)	GPIO_P_MODEL_MODEx(1, mode)
+
+#define GPIO_P_MODEL_MODE2_MASK		GPIO_P_MODEL_MODEx_MASK(2)
+#define GPIO_P_MODEL_MODE2(mode)	GPIO_P_MODEL_MODEx(2, mode)
+
+#define GPIO_P_MODEL_MODE3_MASK		GPIO_P_MODEL_MODEx_MASK(3)
+#define GPIO_P_MODEL_MODE3(mode)	GPIO_P_MODEL_MODEx(3, mode)
+
+#define GPIO_P_MODEL_MODE4_MASK		GPIO_P_MODEL_MODEx_MASK(4)
+#define GPIO_P_MODEL_MODE4(mode)	GPIO_P_MODEL_MODEx(4, mode)
+
+#define GPIO_P_MODEL_MODE5_MASK		GPIO_P_MODEL_MODEx_MASK(5)
+#define GPIO_P_MODEL_MODE5(mode)	GPIO_P_MODEL_MODEx(5, mode)
+
+#define GPIO_P_MODEL_MODE6_MASK		GPIO_P_MODEL_MODEx_MASK(6)
+#define GPIO_P_MODEL_MODE6(mode)	GPIO_P_MODEL_MODEx(6, mode)
+
+#define GPIO_P_MODEL_MODE7_MASK		GPIO_P_MODEL_MODEx_MASK(7)
+#define GPIO_P_MODEL_MODE7(mode)	GPIO_P_MODEL_MODEx(7, mode)
```

Comment:
I thinkall of this is gross, no-one's ever going to use the GPIO_P_MODEL_MODE7_MASK macros individually, but whatever, I see this is just how the existing lg code was, so it's fine :)